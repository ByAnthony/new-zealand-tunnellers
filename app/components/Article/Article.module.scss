@use "app/variables.module.scss";

.container {
  @include variables.component-margin;

  margin-top: variables.$l-margin;

  h1 {
    margin-top: 0;
    margin-bottom: 0;
  }

  p {
    margin: variables.$m-margin 0;
  }

  @include variables.tablet {
    margin-top: variables.$xl-margin;
  }

  @include variables.desktop {
    margin-top: variables.$xxl-margin;
  }
}

.links {
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  padding: variables.$base-margin 0;
  gap: variables.$s-margin;

  @include variables.tablet {
    padding: variables.$base-margin 0;
  }

  @include variables.desktop {
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    padding: variables.$l-margin 0;
  }
}

.link-container {
  height: 350px;
  width: 15rem;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;

  @include variables.featured-button-link;

  p {
    text-align: left;
  }

  .link-button {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: flex-start;
    align-content: space-between;
    padding: variables.$m-margin;
    font-size: 1.5rem;
    color: variables.$secondary-color;

    p {
      margin: 0 0 5px 0;
      font-size: 1.1rem;
      color: variables.$secondary-color;

      @include variables.desktop {
        font-size: 1.2rem;
      }
    }

    @include variables.tablet {
      padding: variables.$m-margin;
      font-size: 1.75rem;
    }

    @include variables.desktop {
      padding: variables.$padding-card;
      font-size: 1.8rem;
    }
  }
}
