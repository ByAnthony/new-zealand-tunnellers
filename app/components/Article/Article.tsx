"use client";

import { Content } from "./Content/Content";
import { HowToCite } from "../HowToCite/HowToCite";
import { ArticleNextChapterButton } from "./ArticleNextChapterButton/ArticleNextChapterButton";
import { ArticleNotes } from "./ArticleNotes/ArticleNotes";
import { Title } from "../Title/Title";
import { TopImage } from "./TopImage/TopImage";
import { Chapter } from "../../types/article";
import { today } from "../../utils/helpers/date";

import STYLES from "./Article.module.scss";

type Props = {
  article: Chapter;
};

export function Article({ article }: Props) {
  return (
    <div className={STYLES.container}>
      <div className={STYLES.header}>
        <div className={STYLES.link}>
          <a href="/#history">History</a>
        </div>
        <Title title={article.title} subTitle={article.chapter} />
      </div>
      <TopImage image={article.image[0]} />
      <Content
        imageList={article.image.slice(1)}
        sectionList={article.section}
      />
      <ArticleNextChapterButton chapter={article.next} />
      <ArticleNotes notes={article.notes} />
      <HowToCite title={article.title} today={today} />
    </div>
  );
}
