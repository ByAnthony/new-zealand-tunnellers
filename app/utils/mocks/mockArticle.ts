import { AboutUsArticle } from "../../types/article";

export const mockArticle: AboutUsArticle = {
  id: "about-us",
  title: "Our Story",
  section: [
    { title: "Title 1", text: "This is the first paragraph of our story." },
    { title: "Title 2", text: "This is the second paragraph of our story." },
  ],
  image: [
    {
      file: "our-story.jpg",
      title: null,
      photographer: null,
      reference: null,
      alt: "A picture of our team",
    },
  ],
};
