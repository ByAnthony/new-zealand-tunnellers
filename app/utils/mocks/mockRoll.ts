import { Tunneller } from '@/app/types/roll';
import { mockId } from './mockProfile';

const johnDoe = {
  forename: 'John',
  surname: 'Doe',
};

const alexanderDriscott = {
  forename: 'Alexander',
  surname: 'Driscott',
};

const williamRight = {
  forename: 'William',
  surname: 'Right',
};

const robertLang = {
  forename: 'Robert',
  surname: 'Lang',
};

export const mockDetails: Tunneller = {
  id: mockId,
  name: johnDoe,
  birthDate: '1886',
  deathDate: '1952',
};

export const mockDetailsList: Tunneller[] = [
  mockDetails,
  {
    ...mockDetails,
    id: 1,
    name: alexanderDriscott,
  },
];

export const mockRoll: Record<string, Tunneller[]> = {
  D: [
    mockDetails,
    {
      ...mockDetails,
      id: 1,
      name: alexanderDriscott,
    },
  ],
  L: [
    {
      ...mockDetails,
      id: 2,
      name: robertLang,
    },
  ],
  R: [
    {
      ...mockDetails,
      id: 3,
      name: williamRight,
    },
  ],
};
