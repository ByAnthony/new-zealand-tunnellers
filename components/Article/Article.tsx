"use client";

import Link from "next/link";
import { useEffect } from "react";

import { ArticleNextChapterButton } from "@/components/Article/ArticleNextChapterButton/ArticleNextChapterButton";
import { ArticleNotes } from "@/components/Article/ArticleNotes/ArticleNotes";
import { Content } from "@/components/Article/Content/Content";
import { TopImage } from "@/components/Article/TopImage/TopImage";
import { HowToCite } from "@/components/HowToCite/HowToCite";
import { Title } from "@/components/Title/Title";
import { Chapter } from "@/types/article";

import STYLES from "./Article.module.scss";

type Props = {
  article: Chapter;
};

export function Article({ article }: Props) {
  useEffect(() => {
    localStorage.clear();
  }, []);

  return (
    <div className={STYLES.container}>
      <div className={STYLES.header}>
        <div className={STYLES.link}>
          <Link href="/#history">History</Link>
        </div>
        <Title title={article.title} subTitle={article.chapter} />
      </div>
      <TopImage image={article.image[0]} />
      <Content
        imageList={article.image.slice(1)}
        sectionList={article.section}
      />
      <ArticleNextChapterButton chapter={article.next} />
      <ArticleNotes notes={article.notes} />
      <HowToCite title={article.title} />
    </div>
  );
}
