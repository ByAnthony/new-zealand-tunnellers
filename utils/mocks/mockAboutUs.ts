import { AboutUsArticle } from "@/types/article";

import {
  mockSectionOne,
  mockSectionTwo,
  mockTitle,
  mockTopImage,
} from "./mockArticle";

export const mockAboutUs: AboutUsArticle = {
  id: "my-awesome-article",
  title: mockTitle,
  section: [mockSectionOne, mockSectionTwo],
  image: [mockTopImage],
};
