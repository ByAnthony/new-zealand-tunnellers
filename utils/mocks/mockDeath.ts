import { Death } from "@/types/tunneller";

import { mockDeath } from "./mockTunneller";

export const mockAfterWarDeath: Death = {
  ...mockDeath,
  warInjuriesDeathAfterWar: true,
};
