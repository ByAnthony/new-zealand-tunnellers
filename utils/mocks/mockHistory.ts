import { HistoryChapter } from "@/types/homepage";

export const mockHistory: HistoryChapter[] = [
  {
    id: "where-all-started",
    chapter: 1,
    title: "Where All\\Started",
    image: "where-all-started.jpg",
  },
  {
    id: "the-journey-ahead",
    chapter: 2,
    title: "The Journey\\Ahead",
    image: "the-journey-ahead.jpg",
  },
  {
    id: "discovering-the-war",
    chapter: 3,
    title: "Discovering\\The War",
    image: "discovering-the-war.jpg",
  },
  {
    id: "below-the-surface",
    chapter: 4,
    title: "Below\\The Surface",
    image: "below-the-surface.jpg",
  },
  {
    id: "back-to-the-surface",
    chapter: 5,
    title: "Back\\To The Surface",
    image: "back-to-the-surface.jpg",
  },
  {
    id: "reconversion-of-activities",
    chapter: 6,
    title: "Reconversion\\Of Activities",
    image: "reconversion-of-activities.jpg",
  },
  {
    id: "returnin-home",
    chapter: 7,
    title: "Returning\\Home",
    image: "returning-home.jpg",
  },
];
