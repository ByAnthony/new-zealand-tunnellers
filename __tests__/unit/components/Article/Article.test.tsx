import { render, screen } from "@testing-library/react";
import { Article } from "@/components/Article/Article";
import { mockArticle } from "@/utils/mocks/mockArticle";

jest.useFakeTimers().setSystemTime(new Date("2023-05-04"));

describe("Article", () => {
  test("matches the snapshot", () => {
    const { asFragment } = render(<Article article={mockArticle} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders the component correctly", () => {
    render(<Article article={mockArticle} />);

    const historyLink = screen.getByText("History");
    expect(historyLink).toHaveAttribute("href", "/#history");

    const titleLineOne = screen.getByText("My Awesome");
    expect(titleLineOne).toHaveClass("title-line-1");

    const titleLineTwo = screen.getByText("Article Title");
    expect(titleLineTwo).toHaveClass("title-line-2");

    expect(screen.getByText("Section Title 1")).toBeInTheDocument();
    expect(screen.getByText("Section text one")).toBeInTheDocument();

    const images = screen.getAllByRole("img");
    expect(images).toHaveLength(2);
    expect(images[0]).toHaveAttribute("alt", "Accessible alt text");
    expect(images[0]).toHaveAttribute(
      "src",
      "/_next/image?url=%2Fimages%2Fhistory%2Fimg-123.png&w=1920&q=75",
    );
    expect(images[1]).toHaveAttribute("alt", "Accessible alt text");
    expect(images[1]).toHaveAttribute(
      "src",
      "/_next/image?url=%2Fimages%2Fhistory%2Fimg-123.png&w=1920&q=75",
    );
    expect(screen.getByText("An Amazing Photo")).toBeInTheDocument();
    expect(
      screen.getByText("The French Photographer Doisneau"),
    ).toBeInTheDocument();
    expect(
      screen.getByText("Where the photo is preserved"),
    ).toBeInTheDocument();

    expect(screen.getByText("Section Title 2")).toBeInTheDocument();
    expect(screen.getByText("Section text two")).toBeInTheDocument();

    const nextButton = screen.getByRole("link", {
      name: "Go to Chapter 3: Next Chapter",
    });
    expect(nextButton).toHaveAttribute(
      "href",
      "/history/my-path-to-next-chapter/",
    );
    expect(screen.getByText("Chapter 3")).toBeInTheDocument();
    expect(screen.getByText("Next Chapter")).toBeInTheDocument();

    expect(screen.getByRole("heading", { name: "Notes" })).toBeInTheDocument();
    const noteOne = screen.getByText("1.");
    expect(noteOne).toHaveAttribute("href", "#reference_1");
    expect(noteOne).toHaveAttribute("id", "footnote_1");
    const noteTwo = screen.getByText("2.");
    expect(noteTwo).toHaveAttribute("href", "#reference_2");
    expect(noteTwo).toHaveAttribute("id", "footnote_2");

    expect(
      screen.getByRole("heading", { name: "How to cite this page" }),
    ).toBeInTheDocument();
    expect(screen.getByText("My Awesome Article Title")).toBeInTheDocument();
    expect(
      screen.getByText(/history\/my-awesome-article-title./),
    ).toBeInTheDocument();
    expect(screen.getByText(/Accessed: 4 May 2023./)).toBeInTheDocument();
  });

  test("does not render the next chapter link when null", () => {
    render(<Article article={{ ...mockArticle, next: null }} />);

    expect(
      screen.queryByRole("link", {
        name: "Go to Chapter 3: Next Chapter",
      }),
    ).toBeNull();
  });
});
