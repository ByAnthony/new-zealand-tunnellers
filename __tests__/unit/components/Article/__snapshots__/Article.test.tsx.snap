// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="header"
    >
      <div
        class="link"
      >
        <a
          href="/#history"
        >
          History
        </a>
      </div>
      <div
        class="main-title"
      >
        <h1>
          <span
            class="title-line-1"
          >
            My Awesome
          </span>
          <span
            class="title-line-2"
          >
            Article Title
          </span>
        </h1>
      </div>
      <h2
        class="title-line-3"
      >
        Chapter 1
      </h2>
    </div>
    <div
      class="image-container"
    >
      <img
        alt="Accessible alt text"
        class="image"
        data-nimg="1"
        decoding="async"
        height="575"
        src="/_next/image?url=%2Fimages%2Fhistory%2Fimg-123.png&w=1920&q=75"
        srcset="/_next/image?url=%2Fimages%2Fhistory%2Fimg-123.png&w=828&q=75 1x, /_next/image?url=%2Fimages%2Fhistory%2Fimg-123.png&w=1920&q=75 2x"
        style="color: transparent;"
        width="800"
      />
    </div>
    <div>
      <div
        class="paragraph"
      >
        <h2>
          Section Title 1
        </h2>
        <p>
          <span>
            Section text one
          </span>
        </p>
      </div>
      <div
        class="image-container"
      >
        <img
          alt="Accessible alt text"
          class="image"
          data-nimg="1"
          decoding="async"
          height="575"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2Fhistory%2Fimg-123.png&w=1920&q=75"
          srcset="/_next/image?url=%2Fimages%2Fhistory%2Fimg-123.png&w=828&q=75 1x, /_next/image?url=%2Fimages%2Fhistory%2Fimg-123.png&w=1920&q=75 2x"
          style="color: transparent;"
          width="800"
        />
      </div>
      <div
        class="image-legend"
      >
        <div
          class="title"
        >
          An Amazing Photo
        </div>
        <div
          class="captions"
        >
          The French Photographer Doisneau
        </div>
        <div
          class="reference"
        >
          Where the photo is preserved
        </div>
      </div>
    </div>
    <div>
      <div
        class="paragraph"
      >
        <h2>
          Section Title 2
        </h2>
        <p>
          <span>
            Section text two
          </span>
        </p>
      </div>
    </div>
    <div
      class="button-chapter-container"
    >
      <a
        aria-label="Go to Chapter 3: Next Chapter"
        class="button-chapter"
        href="/history/my-path-to-next-chapter"
      >
        <div>
          <p
            class="chapter"
          >
            Chapter 3
          </p>
          <span>
            Next Chapter
          </span>
        </div>
        <div
          class="arrow"
        >
          →
        </div>
      </a>
    </div>
    <div
      class="notes"
    >
      <h2>
        Notes
      </h2>
      <p>
        <a
          aria-label="Jump back to the footnote 1"
          class="anchor-link"
          href="#reference_1"
          id="footnote_1"
        >
          1.
        </a>
        <span>
           Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </span>
      </p>
      <p>
        <a
          aria-label="Jump back to the footnote 2"
          class="anchor-link"
          href="#reference_2"
          id="footnote_2"
        >
          2.
        </a>
        <span>
           Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </span>
      </p>
    </div>
    <div
      class="howtocite"
    >
      <h3>
        How to cite this page
        <button
          class="copy-paste"
        >
          <img
            alt="Copy to clipboard"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/_next/image?url=%2Fcopy.png&w=32&q=75"
            srcset="/_next/image?url=%2Fcopy.png&w=16&q=75 1x, /_next/image?url=%2Fcopy.png&w=32&q=75 2x"
            style="color: transparent;"
            width="13"
          />
        </button>
      </h3>
      <p>
        Anthony Byledbal, “
        <span>
          My Awesome Article Title
        </span>
        “,
        <em>
           New Zealand Tunnellers Website
        </em>
        , 2023 (2009), Accessed: 4 May 2023. 
        <span>
          URL: www.
          <wbr />
          nztunnellers
          <wbr />
          .com/
          <wbr />
          history/
          <wbr />
          my-awesome-article-title
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
